server/pulp/server/controllers/repository.py

```diff
@@ -235,6 +235,12 @@ def find_repo_content_units(
             if unit_fields:
                 qs = qs.only(*unit_fields)
 
+            # for performace reasons, do server-side counting of units the will be returned in
+            # result before we do actual query to db
+            # if expected count of units in result is zero, continue to next chunk of unit_ids
+            if qs.count() == 0:
```

Comment:
The logic here looks right, no concerns from "is it functionally correct" point of view.

Can I ask though whether the performance impact is still just a theory or whether it was verified by testing?  It's really surprising to me if the mongo performance of doing a count with no results, vs doing a find with no results, has a notable difference.  It seems like the DB should be performing almost exactly the same work in both cases.

I tried some testing myself against one of the actual DBs involved (QA env), using scripts: https://gist.github.com/rohanpm/1652c648a135a22af34547f7faf19e05

It doesn't simulate the queries here exactly, but it deals with same number of ids in "$in" and uses a query which will return no results.  Timing it shows up pretty much as I expected, no notable difference between the two.

That test of course is not definitive, but if this hasn't been tested I'm a little skeptical about it.

Reply:
My theory is that degradation happens due to instantiation of qs for every chunk - which I think creates new mongo cursor every time. If my theory is correct I think this won't help to resolve the problem

Reply:
I tested the solution with partial prod db on my laptop with pub push-advisory fir e2e repos.
When I debugged where exactly execution waits incredible amount of time it was exactly on `for unit in qs:` 
With skipping on count==0, performance was almost back to normal.
It is true that qs is instantiated for every chunk but actual query is happening during the iteration over qs.
I'll try to come up with some reproducer script, it looks it needs more realistic context for replicating the problem.

Reply:
Based on the testing done elsewhere (RHELDST-2771), although this is unintuitive and might point to some deficiency in mongo or one of the client libraries, it really is giving a major performance benefit.  Apparently, doing a search which finds nothing can be a lot slower than doing a count which finds nothing.

This PR gets a :+1: from me as we have evidence that it will significantly improve the performance.